File: 3219-minimum-cost-for-cutting-cake-ii/3219-minimum-cost-for-cutting-cake-ii.js
/**
 * @param {number} m
 * @param {number} n
 * @param {number[]} horizontalCut
 * @param {number[]} verticalCut
 * @return {number}
 */
var minimumCost = function (m, n, horCut, verCut) {
  let totalCost = 0;
  const cuts = [1, 1];
  [...horCut.map(v => [v, 1]), ...verCut.map(v => [v, 0])].sort((x, y) => y[0] - x[0])
    .forEach(([cost, dir]) => (totalCost += cost * cuts[dir], cuts[1 - dir]++));
  return totalCost;
};

File: 3219-minimum-cost-for-cutting-cake-ii/3219-minimum-cost-for-cutting-cake-ii.java
class Solution {
    public long minimumCost(int m, int n, int[] horizontalCut, int[] verticalCut) {
        Arrays.sort(horizontalCut);
        Arrays.sort(verticalCut);
        
        long cost = 0;
        int i=horizontalCut.length-1, j=verticalCut.length-1;
        int hCuts = 1, vCuts = 1;
        
        while(i>=0 && j>=0){
            if(horizontalCut[i] > verticalCut[j]){
                cost += horizontalCut[i] * vCuts;
                hCuts++; i--;
            }
            else{
                cost += verticalCut[j] * hCuts;
                vCuts++; j--;
            }
        }
        while(i>=0){
            cost += horizontalCut[i] * vCuts;
            i--;
        }
        while(j>=0){
            cost += verticalCut[j] * hCuts;
            j--;
        }
        return cost;
    }
}